app/master/build.py

```diff
@@ -309,6 +288,41 @@ def _on_enter_error_state(self, event):
         self._error_message = getattr(event, 'error_msg', default_error_msg)
         self._logger.warning('Build {} failed: {}', self.build_id(), self._error_message)
 
+    def _on_enter_preparing_state(self, event):
+        """
+        Prepare the build by atomization and subjobs creation.
+        This method is triggered by a state machine transition to the PREPARING state.
+        :param event: The Fysom event object
+        """
+        self._logger.info('Fetching project for build {}.', self._build_id)
+        self.project_type.fetch_project()
+        self._logger.info('Successfully fetched project for build {}.', self._build_id)
+
+        job_config = self.project_type.job_config()
+        if job_config is None:
+            raise RuntimeError('Build failed while trying to parse clusterrunner.yaml.')
+
+        subjob_calculator = getattr(event, 'subjob_calculator', None)
```

Comment:
`subjob_calculator` comes from `BuildRequestHandler`. Personally I will not prefer it to be part of callback `build._on_enter_preparing_state`. Also, I am not sure but, to move it we may need to make at multiple places.